vyper/semantics/analysis/imports.py

```diff
@@ -0,0 +1,328 @@
+    InputBundle,
+    PathLike,
+)
+from vyper.exceptions import (
+    CompilerPanic,
+    DuplicateImport,
+    ImportCycle,
+    ModuleNotFound,
+    StructureException,
+)
+from vyper.semantics.analysis.base import ImportInfo
+from vyper.utils import safe_relpath, sha256sum
+
+"""
+collect import statements and validate the import graph.
+this module is separated into its own pass so that we can resolve the import
+graph quickly (without doing semantic analysis) and for cleanliness, to
+segregate the I/O portion of semantic analysis into its own pass.
+"""
+
+
+@dataclass
+class _ImportGraph:
+    # the current path in the import graph traversal
+    _path: list[vy_ast.Module] = field(default_factory=list)
+
+    # stack of dicts, each item in the stack is a dict keeping
+    # track of imports in the current module
+    _imports: list[dict] = field(default_factory=list)
+
+    @property
+    def imported_modules(self):
+        return self._imports[-1]
+
+    @property
+    def current_module(self):
+        return self._path[-1]
+
+    def push_path(self, module_ast: vy_ast.Module) -> None:
+        if module_ast in self._path:
+            cycle = self._path + [module_ast]
+            raise ImportCycle(" imports ".join(f'"{t.path}"' for t in cycle))
+
+        self._path.append(module_ast)
+        self._imports.append({})
+
+    def pop_path(self, expected: vy_ast.Module) -> None:
+        popped = self._path.pop()
+        if expected != popped:
+            raise CompilerPanic("unreachable")
+        self._imports.pop()
+
+    @contextlib.contextmanager
+    def enter_path(self, module_ast: vy_ast.Module) -> Iterator[None]:
+        self.push_path(module_ast)
+        try:
+            yield
+        finally:
+            self.pop_path(module_ast)
+
+
+class ImportAnalyzer:
+    def __init__(self, input_bundle: InputBundle, graph: _ImportGraph):
+        self.input_bundle = input_bundle
+        self.graph = graph
+        self._ast_of: dict[int, vy_ast.Module] = {}
+
+        self.seen: set[int] = set()
+
+        self.integrity_sum = None
+
+    def resolve_imports(self, module_ast: vy_ast.Module):
+        self._resolve_imports_r(module_ast)
+        self.integrity_sum = self._calculate_integrity_sum_r(module_ast)
+
+    def _calculate_integrity_sum_r(self, module_ast: vy_ast.Module):
+        acc = [sha256sum(module_ast.full_source_code)]
+        for s in module_ast.get_children((vy_ast.Import, vy_ast.ImportFrom)):
+            info = s._metadata["import_info"]
+
+            if info.compiler_input.path.suffix in (".vyi", ".json"):
+                # NOTE: this needs to be redone if interfaces can import other interfaces
+                acc.append(info.compiler_input.sha256sum)
+            else:
+                acc.append(self._calculate_integrity_sum_r(info.parsed))
+
+        return sha256sum("".join(acc))
+
+    def _resolve_imports_r(self, module_ast: vy_ast.Module):
+        if id(module_ast) in self.seen:
```

Comment:
a bit hard to read bc one immediately asks why this doesn't interfere with cycle detection - maybe add a comment?

Reply:
because it gets added to seen after the recursion (on line 110)